apis/Google.Cloud.Spanner.V1/Google.Cloud.Spanner.V1/SessionPool.cs

```diff
@@ -0,0 +1,311 @@
+                lock (s_priorityListSync)
+                {
+                    targetPool.ReleaseSessionToPool(client, session);
+                    if (CurrentPooledSessions > MaximumPooledSessions)
+                    {
+                        var evictionPool = PriorityList.First().Value;
+                        evictionClient = evictionPool.Key.Client;
+                        evictionSession = evictionPool.AcquireEvictionCandidate();
+                        ReSortMru(evictionPool);
+                    }
+                    ReSortMru(targetPool);
+                }
+                if (evictionSession != null)
+                {
+                    Task.Run(() => EvictSession(evictionClient, evictionSession));
+                }
+            }
+        }
+
+        private static async Task EvictSession(SpannerClient evictionClient, Session evictionSession)
+        {
+            await evictionSession.RemoveFromTransactionPool().ConfigureAwait(false);
+            await evictionClient.DeleteSessionAsync(evictionSession.SessionName);
+        }
+
+        /// <summary>
+        /// Called to close a session and release it without putting it back into the pool.
+        /// </summary>
+        /// <param name="session"></param>
+        /// <returns></returns>
+        public static async Task Close(this Session session)
+        {
+            session.ThrowIfNull(nameof(session));
+
+            SessionPoolKey result;
+            if (s_sessionsInUse.TryRemove(session, out result))
+            {
+                LogSessionsInUse();
+                SignalAnyWaitingRequests();
+                await result.Client.DeleteSessionAsync(session.SessionName).ConfigureAwait(false);
+                return;
+            }
+            throw new InvalidOperationException("Close Session was not able to locate the provided session.");
+        }
+
+        private static void ReSortMru(SessionPoolImpl targetPool)
+        {
+            lock (s_priorityListSync)
+            {
+                if (PriorityList.Count > 1)
+                {
+                    PriorityList.Remove(targetPool);
+                    PriorityList.Add(targetPool, targetPool);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Releases all pooled sessions and frees resources on the server.
+        /// </summary>
+        /// <returns></returns>
+        public static Task ReleaseAll()
+        {
+            var entries = GlobalPool.Values;
+            GlobalPool.Clear(); // ReleaseAll should not be called while other operations are starting.
+            return Task.WhenAll(entries.Select(sessionpool => sessionpool.ReleaseAllImpl()).ToArray());
+        }
+
+        /// <summary>
+        /// The current number of active sessions in use by the application.
+        /// </summary>
+        public static int CurrentActiveSessions => s_sessionsInUse.Count + s_sessionsCreating;
```

Comment:
actually I just made added the monitor here as well, because its basically a no-op and makes the code safer long term if someone calls this outside a lock.